noir-r1cs/src/compiler.rs

```diff
@@ -116,19 +119,31 @@ impl R1CS {
     pub fn add_assert_zero(&mut self, expr: &Expression<FieldElement>) {
         // Create individual constraints for all the multiplication terms and collect
         // their outputs
-        let mut linear = expr
-            .mul_terms
-            .iter()
-            .map(|term| {
-                let a = self.map_witness(term.1);
-                let b = self.map_witness(term.2);
-                let c = self.new_witness();
-                self.add_constraint(&[(FieldElement::one(), a)], &[(FieldElement::one(), b)], &[
-                    (FieldElement::one(), c),
-                ]);
-                (term.0, c)
-            })
-            .collect::<Vec<_>>();
+        let mut linear = vec![];
+        let mut a = vec![];
+        let mut b = vec![];
+        if expr.mul_terms.len() > 1 {
+            linear = expr
+                .mul_terms
+                .iter()
+                .map(|term| {
+                    let a = self.map_witness(term.1);
+                    let b = self.map_witness(term.2);
+                    let c = self.new_witness();
+                    self.add_constraint(
+                        &[(FieldElement::one(), a)],
+                        &[(FieldElement::one(), b)],
+                        &[(FieldElement::one(), c)],
+                    );
+                    (term.0, c)
+                })
+                .collect::<Vec<_>>();
+        } else if expr.mul_terms.len() == 1 {
+            // no need to create intermediate multiplication constraints
+            let term = expr.mul_terms[0];
+            a = vec![(FieldElement::one(), self.map_witness(term.1))];
+            b = vec![(FieldElement::one(), self.map_witness(term.2))];
+        }
```

Comment:
There is a more general result: The last mul term never needs an intermediate constraint (even if there are multiple mul-terms).